controller/volume_controller.go

```diff
@@ -848,18 +848,40 @@ func (vc *VolumeController) cleanupExtraHealthyReplicas(v *longhorn.Volume, e *l
 }
 
 func (vc *VolumeController) cleanupEvictionRequestedReplicas(v *longhorn.Volume, rs map[string]*longhorn.Replica) (bool, error) {
+	log := getLoggerForVolume(vc.logger, v)
+
+	// If there is no non-evicting healthy replica,
+	// Longhorn should retain one evicting healthy replica.
+	hasNonEvictingHealthyReplica := false
+	evictingHealthyReplica := ""
 	for _, r := range rs {
-		if r.Status.EvictionRequested {
-			if err := vc.deleteReplica(r, rs); err != nil {
-				vc.eventRecorder.Eventf(v, v1.EventTypeWarning,
-					EventReasonFailedEviction,
-					"volume %v failed to evict replica %v",
-					v.Name, r.Name)
-				return false, err
-			}
-			logrus.Debugf("Evicted replica %v", r.Name)
-			return true, nil
+		if !datastore.IsAvailableHealthyReplica(r) {
+			continue
+		}
+		if !r.Status.EvictionRequested {
+			hasNonEvictingHealthyReplica = true
+			break
 		}
+		evictingHealthyReplica = r.Name
+	}
+
+	for _, r := range rs {
+		if !r.Status.EvictionRequested {
+			continue
+		}
+		if !hasNonEvictingHealthyReplica && r.Name == evictingHealthyReplica {
+			log.Warnf("Cannot evicting replica %v for now since there is no other healthy replica", r.Name)
+			continue
+		}
+		if err := vc.deleteReplica(r, rs); err != nil {
+			vc.eventRecorder.Eventf(v, v1.EventTypeWarning,
+				EventReasonFailedEviction,
+				"volume %v failed to evict replica %v",
+				v.Name, r.Name)
+			return false, err
+		}
+		log.Debugf("Evicted replica %v in disk %v of node %v ", r.Name, r.Spec.DiskID, r.Spec.NodeID)
+		return true, nil
 	}
```

Comment:
Why not just use evictingHealthyReplica as *replica instead of replica name, and then you don't need to have another loop for the replica deletion?

```
	if !hasNonEvictingHealthyReplica {
		log.Warnf("Cannot evicting replica %v for now since there is no other healthy replica", evictingHealthyReplica.Name)
		return false, nil
	}
	  
	if err := vc.deleteReplica(evictingHealthyReplica, rs); err != nil {
		vc.eventRecorder.Eventf(
			v, 
			v1.EventTypeWarning,
			EventReasonFailedEviction,
			"volume %v failed to evict replica %v",
			v.Name, 
			evictingHealthyReplica.Name,
		)
		return false, err
	}
		
	log.Debugf("Evicted replica %v in disk %v of node %v ", r.Name, r.Spec.DiskID, r.Spec.NodeID)
	return true, nil
```

Reply:
If there are multiple evicting healthy replicas, the current implement could remove some of them. But the suggested code, all eviction will be blocked, right?

Reply:
ah, you are right! I missed that, LGTM 🙂